doc/designs/evp_skey.md

```diff
@@ -0,0 +1,183 @@
+EVP_SKEY Design
+===============
+
+We currently presume that symmetric keys are represented as octet buffer. It
+may be not relevant for the situation of hardware-backed providers, such as
+PKCS#11 or TPM.
+
+This is a design of using opaque usage in symmetric encryption.
+
+If the provider allows key exporting, the internal representation of the
+EVP_SKEY is "a bunch of bytes", that can be cached and made unattached to any
+provider, and become useful to any current cipher supporting the symmetric key
+in byte form.
```

Comment:
Although you can detach a key I think an "origin" handle should be preserved and used if an operation ever goes back to the original provider, otherwise this will cause churn in providers that can export but can't "match" back a key to perform operations, and would have to re-import a copy if a trace of what the original handle is not preserved. We see this already with the current EVP-PKEY implementation where OpenSSL aggressively tries to export public keys to the default provider and sometimes has to try to re-import them back to the original provider. This is not only slow but also cause issues because matching a key is not always possible with the APIs available to the provider.
I would like to avoid to repeat the same mistake here.